compiler/optimizations/loopInvariantCodeMotion.cpp

```diff
@@ -910,21 +922,26 @@ static void computeLoopInvariants(std::vector<SymExpr*>& loopInvariants,
       }
     }
 
+    bool isGpuBound = loop->isGpuBound();
     bool mightHaveBeenDeffedElseWhere = false;
     // assume that anything passed in by ref has been changed elsewhere
     // Note that not all things that are passed by ref will have the ref intent
     // flag, and may just be ref variables. This is a known bug, see comments
     // in addVarsToFormals(): flattenFunctions.cpp.
+    //
+    // For GPU-bound loops, relax this restriction and say that changing things
+    // passed to a kernel by reference from underneath the kernel is undefined
+    // behavior.
```

Comment:
Unless I am mistaken while reading the diff, you should be able to wrap this block in an `if (!isGpuBound)` as well. It looks like you are doing at the deepest condition, which seems a bit late potentially causing redundant operations.